File: policy-engine/internal/plugins/ip_whitelist.go
package plugins

import (
	"context"
	"net"
	"strings"

	"github.com/xcloudapim/policy-engine/internal/domain"
)

// IPWhitelistPlugin IP 白名單 / 黑名單
// config keys:
//   mode = "whitelist" | "blacklist"  (default: whitelist)
//   ips  = "1.2.3.4,10.0.0.0/8,192.168.1.0/24"
type IPWhitelistPlugin struct{}

func NewIPWhitelistPlugin() *IPWhitelistPlugin { return &IPWhitelistPlugin{} }

func (p *IPWhitelistPlugin) Type() domain.PolicyType { return domain.PolicyTypeIPWhitelist }

func (p *IPWhitelistPlugin) Validate(config map[string]string) []string {
	var errs []string
	if cfgGet(config, "ips") == "" {
		errs = append(errs, "ips is required")
		return errs
	}
	for _, ipOrCIDR := range parseCSV(cfgGet(config, "ips")) {
		ipOrCIDR = strings.TrimSpace(ipOrCIDR)
		if strings.Contains(ipOrCIDR, "/") {
			if _, _, err := net.ParseCIDR(ipOrCIDR); err != nil {
				errs = append(errs, "invalid CIDR: "+ipOrCIDR)
			}
		} else if net.ParseIP(ipOrCIDR) == nil {
			errs = append(errs, "invalid IP: "+ipOrCIDR)
		}
	}
	return errs
}

func (p *IPWhitelistPlugin) Execute(ctx context.Context, execCtx *domain.ExecContext, config map[string]string) error {
	mode   := cfgGetDefault(config, "mode", "whitelist")
	ipList := parseCSV(cfgGet(config, "ips"))

	// 優先取 X-Forwarded-For（反向代理後面的真實 IP）
	clientIP := execCtx.GetHeader("X-Forwarded-For")
	if clientIP == "" {
		clientIP = execCtx.GetHeader("X-Real-IP")
	}
	if clientIP == "" {
		clientIP = execCtx.RemoteIP
	}
	// X-Forwarded-For 可能是逗號清單，取第一個
	if idx := strings.Index(clientIP, ","); idx != -1 {
		clientIP = strings.TrimSpace(clientIP[:idx])
	}

	parsedIP := net.ParseIP(clientIP)
	if parsedIP == nil {
		execCtx.Abort(400, "invalid client IP address")
		return nil
	}

	matched := p.ipMatches(parsedIP, ipList)

	switch mode {
	case "whitelist":
		if !matched {
			execCtx.Abort(403, "access denied: IP not in whitelist")
		}
	case "blacklist":
		if matched {
			execCtx.Abort(403, "access denied: IP is blacklisted")
		}
	}

	return nil
}

func (p *IPWhitelistPlugin) ipMatches(ip net.IP, list []string) bool {
	for _, entry := range list {
		entry = strings.TrimSpace(entry)
		if entry == "" {
			continue
		}
		if strings.Contains(entry, "/") {
			_, network, err := net.ParseCIDR(entry)
			if err == nil && network.Contains(ip) {
				return true
			}
		} else {
			if net.ParseIP(entry).Equal(ip) {
				return true
			}
		}
	}
	return false
}

File: gateway/src/pipeline/plugins/ip-whitelist.ts
import type { ExecContext, PluginDeps } from '../types.js'
import { abort } from '../types.js'

export async function ipWhitelist(ctx: ExecContext, deps: PluginDeps): Promise<void> {
  const mode   = deps.config['mode'] ?? 'whitelist'
  const ipList = (deps.config['ips'] ?? '').split(',').map((s) => s.trim()).filter(Boolean)

  // 解析客戶端真實 IP
  let clientIp = ctx.requestHeaders['x-forwarded-for'] ?? ctx.requestHeaders['x-real-ip'] ?? ctx.remoteIp
  const commaIdx = clientIp.indexOf(',')
  if (commaIdx >= 0) clientIp = clientIp.slice(0, commaIdx).trim()

  if (!clientIp) {
    abort(ctx, 400, 'invalid client IP address')
    return
  }

  const matched = ipList.some((entry) => ipMatches(clientIp, entry))

  if (mode === 'whitelist' && !matched) {
    abort(ctx, 403, 'access denied: IP not in whitelist')
  } else if (mode === 'blacklist' && matched) {
    abort(ctx, 403, 'access denied: IP is blacklisted')
  }
}

function ipMatches(clientIp: string, entry: string): boolean {
  if (entry.includes('/')) {
    return cidrContains(entry, clientIp)
  }
  return entry === clientIp
}

/** 簡易 IPv4 CIDR 判斷（不依賴外部套件） */
function cidrContains(cidr: string, ip: string): boolean {
  const [range, bits] = cidr.split('/')
  const mask = ~((1 << (32 - parseInt(bits!, 10))) - 1) >>> 0
  const ipInt   = ipToInt(ip)
  const rangeInt = ipToInt(range!)
  if (ipInt === null || rangeInt === null) return false
  return (ipInt & mask) === (rangeInt & mask)
}

function ipToInt(ip: string): number | null {
  const parts = ip.split('.')
  if (parts.length !== 4) return null
  const nums = parts.map(Number)
  if (nums.some((n) => isNaN(n) || n < 0 || n > 255)) return null
  return ((nums[0]! << 24) | (nums[1]! << 16) | (nums[2]! << 8) | nums[3]!) >>> 0
}
